### PR TITLE
Fixes to `gr.Examples`

### PR DESCRIPTION
**Pull request by abidlabs**
This PR fixes some issues with `gr.Examples`:

* The Examples dataset component would not appear correctly if only some of the input components had labels. This is now fixed, with tests to ensure the correct behavior. (Fixes: #1948)
* When examples were read from a log file, they were not loaded correctly because of relative file imports. This is now fixed, with tests to ensure the correct behavior. (Fixes: #1875)
* Additional tests are added for `gr.Examples`

### CODE REVIEW

**Comment by freddyaboulton** [comment]
@abidlabs Is this ready for review? I see it's still Draft and the tests are failing

**Comment by abidlabs** [comment]
Ready now for a review @freddyaboulton, thank you!

File: gradio/examples.py [hunk 1]
```diff
@@ -225,6 +222,7 @@ def load_from_cache(self, example_id: int) -> List[Any]:
         example = examples[example_id + 1]  # +1 to adjust for header
         output = []
         for component, cell in zip(self.outputs, example):
+            print("self.cached_folder", self.cached_folder, "cell", cell)
```
Comment (abidlabs) [inline gradio/examples.py:225]:
```suggestion
```

**Comment by abidlabs** [comment]
As discussed with @freddyaboulton, I don't love the fact that we change directories to read in the examples. Basically the issue is that paths of samples in the log.csv file are relative to the folder containing the log.csv file (call this the _log path_). So we basically have 3 options:

1. Change directories when loading the samples
1. Append the log path to all paths when we read in the csv file (not obvious to know which samples in the log file paths vs. strings, etc.)
1. Change the preprocess_example to also accept in a directory

I feel like changing directories is the best option, but definitely open to discussing (maybe #3 is better)

**Comment by abidlabs** [comment]
We can make #1 more robust by using context managers to wrap the change in directory, e.g. https://dev.to/teckert/changing-directory-with-a-python-context-manager-2bj8

**Comment by abidlabs** [comment]
I've implemented the changing of the working directory with context managers instead, I think it's more readable and robust now.

**Review by freddyaboulton — APPROVED** [review]
Thank you for the fix, using the context manager approach and adding the unit tests @abidlabs ! 

This is good to merge

**Comment by abidlabs** [comment]
Thanks for your review and suggestions @freddyaboulton! Will fix and merge

**Pull request merged**